File: lua/bottle-song.lua
local BottleSong = {}

local numbers = { "One", "Two", "Three", "Four", "Five", "Six", "Seven", "Eight", "Nine", "Ten" }

numbers[0] = "no"

local function s(bottles)
    if bottles == 1 then
        return "bottle"
    else
        return "bottles"
    end
end

local function verse(bottles)
    return {
        numbers[bottles] .. " green " .. s(bottles) .. " hanging on the wall,\n",
        numbers[bottles] .. " green " .. s(bottles) .. " hanging on the wall,\n",
        "And if one green bottle should accidentally fall,\n",
        "There'll be " .. numbers[bottles - 1]:lower() .. " green " .. s(bottles - 1) .. " hanging on the wall.\n",
    }
end

function BottleSong.recite(start_bottles, take_down)
    local t = {}
    for _ = take_down, 1, -1 do
        local o = verse(start_bottles)
        if #t > 0 then
            table.insert(t, "\n")
        end
        for _, v in ipairs(o) do
            table.insert(t, v)
        end

        start_bottles = start_bottles - 1
    end

    return table.concat(t)
end

return BottleSong

File: go/bottle_song.go
package bottlesong

import (
	"fmt"
	"strings"
)

var nums = []string{"no", "one", "two", "three", "four", "five", "six", "seven", "eight", "nine", "ten"}

func aux(n int) string {
	if n == 1 {
		return "bottle"
	}
	return "bottles"
}

func verses(n int) []string {
	first := fmt.Sprintf("%s green %s hanging on the wall,", nums[n], aux(n))
	first = strings.ToUpper(first[:1]) + first[1:]
	verse := []string{
		first,
		first,
		"And if one green bottle should accidentally fall,",
		fmt.Sprintf("There'll be %s green %s hanging on the wall.", nums[n-1], aux(n-1)),
	}
	return verse
}

func Recite(startBottles, takeDown int) []string {
	arr := []string{}
	for i := 0; i < takeDown; i++ {
		if len(arr) > 0 {
			arr = append(arr, "")
		}
		arr = append(arr, verses(startBottles)...)
		startBottles--
	}
	return arr
}
